### PR TITLE
Entity whitelist for raypicking and whiteboard painting example

### PR DESCRIPTION
**Pull request by ericrius1**
This PR adds support for whitelisting entities to be raypicked. If the user provides an array of one or more entityIDs to as the 3rd argument of Entities.findRayIntersection() call, only entities contained in that array will be whitelisted and picked against. Otherwise they will be ignored.

This helps with raypicking performance when user knows there are certain items they want to pick against, and other entities in the scene can be ignored. For example, create a whiteboard in a complex scene such as toybox by runnning whiteboardSpawner.js. Now, in whiteboardEntityScript.js, uncomment line 71 and comment in line 73. The script will now pick against all entities in the scene, including complex models with many thousands of unsorted triangles, twice each frame (most likely at 60 fps- yikes!) as opposed to only picking the whiteboard. A noticeable increase in judder in the Rift becomes apparent.

The whiteboard scripts included in the PR comprise an app, that, when whiteboardSpawner is run, place a whiteboard in front of the client's avatar which ran the script. Any user can now approach the whiteboard and draw on it using their hydras. Users can switch colors by clicking or hydra grabbing the color boxes lining the top of the whiteboard. 

To test this script, run 
https://rawgit.com/ericrius1/hifi/whiteboard/examples/painting/whiteboard/whiteboardSpawner.js
 and walk up to the whiteboard and start painting with either hydra by holding z-trigger. (Make sure to test with this PR build so whitelist is honored)

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: db5f7cac8dc5e7f6de489c9fdcc63a2e764c48a9 available at: https://gist.github.com/3bfb2ed021f6f48baa75<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: eb7b720945f385cdc318b9b67aed8dbdc5546877 available at: https://gist.github.com/3bfb2ed021f6f48baa75<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0b92bc11f6a47068cbc8814cd651138f824b473e available at: https://gist.github.com/3bfb2ed021f6f48baa75<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.dmg

**Comment by imgntn** [comment]
@ericrius1  a couple of first observations:

1. cool
2. sometimes i can't paint over previous colors.  i.e. if i paint green, then pick pink, when i paint over it stays green for some of the stroke
3. if i'm too close to the board i can't paint (my beam just locks on to the middle of the board)
 --> maybe try putting a collision hull on the board that is bigger than the board so it would keep people a reasonable distance away
4. my hand grab beams go right through the board --  they should probably stop
5. we need an eraser -- just make a color block that is the same color as the board.  you don't always want to erase the whole board
6. if i use right trigger to paint and hold it down, my reticle disappears if i click left trigger and vice versa
7. the back of the whiteboard looks just like the front, except you can't write on it.  if you spawn the whiteboard away from a wall (in the middle of a room) it can be confusing which way to approach it from.  either both sides paintable or make the non-paintable side look less paintable

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 1]
```diff
@@ -414,15 +416,20 @@ void RayToEntityIntersectionResultFromScriptValue(const QScriptValue& object, Ra
     QString faceName = object.property("face").toVariant().toString();
     if (faceName == "MIN_X_FACE") {
         value.face = MIN_X_FACE;
-    } else if (faceName == "MAX_X_FACE") {
+    }
```
Comment (ZappoMan) [inline libraries/entities/src/EntityScriptingInterface.cpp]: coding standard - brace placement `} else if() {`

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 2]
```diff
@@ -414,15 +416,20 @@ void RayToEntityIntersectionResultFromScriptValue(const QScriptValue& object, Ra
...
+    else if (faceName == "MAX_X_FACE") {
```
Comment (ZappoMan) [inline libraries/entities/src/EntityScriptingInterface.cpp]: coding standard - brace placement `} else if() {`

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 3]
```diff
@@ -414,15 +416,20 @@ void RayToEntityIntersectionResultFromScriptValue(const QScriptValue& object, Ra
...
         value.face = MAX_X_FACE;
-    } else if (faceName == "MIN_Y_FACE") {
+    }
+    else if (faceName == "MIN_Y_FACE") {
```
Comment (ZappoMan) [inline libraries/entities/src/EntityScriptingInterface.cpp]: coding standard - brace placement `} else if() {`

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 4]
```diff
@@ -414,15 +416,20 @@ void RayToEntityIntersectionResultFromScriptValue(const QScriptValue& object, Ra
...
         value.face = MIN_Y_FACE;
-    } else if (faceName == "MAX_Y_FACE") {
+    }
+    else if (faceName == "MAX_Y_FACE") {
```
Comment (ZappoMan) [inline libraries/entities/src/EntityScriptingInterface.cpp]: coding standard - brace placement `} else if() {`

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 5]
```diff
@@ -414,15 +416,20 @@ void RayToEntityIntersectionResultFromScriptValue(const QScriptValue& object, Ra
...
         value.face = MAX_Y_FACE;
-    } else if (faceName == "MIN_Z_FACE") {
+    }
+    else if (faceName == "MIN_Z_FACE") {
```
Comment (ZappoMan) [inline libraries/entities/src/EntityScriptingInterface.cpp]: coding standard - brace placement `} else if() {`

File: libraries/entities/src/EntityScriptingInterface.cpp [hunk 6]
```diff
@@ -414,15 +416,20 @@ void RayToEntityIntersectionResultFromScriptValue(const QScriptValue& object, Ra
...
         value.face = MIN_Z_FACE;
-    } else {
+    }
+    else {
```
Comment (ZappoMan) [inline libraries/entities/src/EntityScriptingInterface.cpp]: coding standard - brace placement `} else if() {`

File: libraries/entities/src/EntityTreeElement.cpp [hunk 7]
```diff
@@ -495,12 +495,25 @@ bool EntityTreeElement::bestFitBounds(const glm::vec3& minPoint, const glm::vec3
...
+                                    const QVector<QUuid>& entityIdsToInclude, void** intersectedObject, bool precisionPicking, float distanceToElementCube) {
 
     // only called if we do intersect our bounding cube, but find if we actually intersect with entities...
     int entityNumber = 0;
     bool somethingIntersected = false;
     forEachEntity([&](EntityItemPointer entity) {
+        if (entityIdsToInclude.size() > 0) {
+            bool entityInWhiteList = false;
+            // We only want to search whitelist if there is one, otherwise everything except blacklisted items are valid
+            for (int i = 0; i < entityIdsToInclude.size(); i++) {
```
Comment (ZappoMan) [inline libraries/entities/src/EntityTreeElement.cpp]: Why not use `QVector::contains()`?

```
if (entityIdsToInclude.contains(entity->getID())) {
    entityInWhiteList = true;
}
```

File: libraries/octree/src/Octree.cpp [hunk 8]
```diff
@@ -721,9 +722,9 @@ bool findRayIntersectionOp(OctreeElementPointer element, void* extraData) {
 
 bool Octree::findRayIntersection(const glm::vec3& origin, const glm::vec3& direction,
                                     OctreeElementPointer& element, float& distance, 
-                                    BoxFace& face, glm::vec3& surfaceNormal, void** intersectedObject,
+                                    BoxFace& face, glm::vec3& surfaceNormal, const QVector<QUuid>& entityIdsToInclude, void** intersectedObject,
                                     Octree::lockType lockType, bool* accurateResult, bool precisionPicking) {
-    RayArgs args = { origin, direction, element, distance, face, surfaceNormal, intersectedObject, false, precisionPicking};
+    RayArgs args = { origin, direction, element, distance, face, surfaceNormal, entityIdsToInclude, intersectedObject, false, precisionPicking};
```
Comment (ZappoMan) [inline libraries/octree/src/Octree.cpp]: we shouldn't add the idea of "entityIDs" to the Octree code... instead move this code to EntityTree

Comment (ericrius1) [inline libraries/octree/src/Octree.cpp]: Are you saying I should I move the entity RayArgs class to the EntityTree from here?

**Comment by hifi-gustavo** [comment]
Build console output for commit: cfca0fe8c02bbd06de7dce7965675dd368bbaabc available at: https://gist.github.com/3bfb2ed021f6f48baa75<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4c709cba3551763195a6af37a3e6ce105b9cd09c available at: https://gist.github.com/3bfb2ed021f6f48baa75<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5082ac458316b6d0d2b548886459e35015db9605 available at: https://gist.github.com/3bfb2ed021f6f48baa75<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 87439a049e90ef3792f4bc3d9076458e9bbcf7ce available at: https://gist.github.com/3bfb2ed021f6f48baa75<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3b59bffb79ae88caa6e703e52cc6120132729379 available at: https://gist.github.com/3bfb2ed021f6f48baa75<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: dc6875f5b582d648b35bd94f14dbf165d608e424 available at: https://gist.github.com/3bfb2ed021f6f48baa75<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e8b6b32572afc075b9c21a44b15f4f19d3f6b6c4 available at: https://gist.github.com/3bfb2ed021f6f48baa75<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 97fe0170cbc5a20aa6d89dd83176082cbbe3e67d available at: https://gist.github.com/3bfb2ed021f6f48baa75<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6079.dmg

**Pull request merged**